horovod/common/controller.cc

```diff
@@ -98,6 +100,12 @@ ResponseList Controller::ComputeResponseList(std::atomic_bool& shut_down,
     }
   }
 
+  if (state.joined && response_cache_.capacity() > 0) {
+    for (uint32_t bit : response_cache_.list_all_bits()) {
+      cache_coordinator.record_hit(bit);
```

Comment:
Noticed a subtle issue with this logic here to set all cache bits if `state.joined = true`. On the cycle where the last worker submits a join, all workers will signal that all responses in the cache are hit, and will proceed to retrieve process every response in the cache. The extra status flag may be useful for this case, except you'd want to catch if all workers are in a join state, and if so, erase all cache hits.

Reply:
> Noticed a subtle issue with this logic here to set all cache bits if state.joined = true. On the cycle where the last worker submits a join, all workers will signal that all responses in the cache are hit, and will proceed to retrieve process every response in the cache. The extra status flag may be useful for this case, except you'd want to catch if all workers are in a join state, and if so, erase all cache hits.

On the cycle where the last worker submits a join, `need_communication` is true, so responses from hits are generated on the coordinator, which knows the `joined_size` and can just not generate these responses.